sections/advanced/email-testing.md

```diff
@@ -0,0 +1,213 @@
+However, these approaches can be problematic; for example non-existing emails can prompt bouncing emails to your cloud service and that can be a headache. If you want to avoid such issues and check real email content in automation, email services provide value. 
+
+Email services can also provide cost savings in text execution time by receiving emails faster, tests running quicker in the pipeline with less CI resources being consumed and less time waiting for tests to finish. If you are running 1000 pipelines a year, and save 3-4 seconds per pipeline execution, the email service can be already paying for its annual subscription just by providing extra speed.
+
+
+### Achieving Stateless tests with unique emails
+
+If in every test execution a new, unique user was used and the emails to this unique user could be verified in isolation, it would be possible to achieve a stateless test. The only side effect would be to the email service inbox, but if the test only checked emails by reference and cleaned up after itself, the email service mailbox would not be impacted.
+
+This is easy to achieve with Mailosaur, here are two approaches to this ; [Mailosaur's Node package](https://www.npmjs.com/package/mailosaur) or our own util with `faker.js`.
+
+```javascript
+// at cypress/plugins/mailosaur-tasks.js
+
+// generates a random email address
+// sample output:   ojh788.<serverId>@mailosaur.io
+const createEmail = () => mailosaurClient
+  .servers
+  .generateEmailAddress(envVars.MAILOSAUR_SERVERID);
+);
+
+// our custom function at a helper file or commands file. The only difference is the defined prefixed name.
+// sample output:  fakerJsName.<serverId>@mailosaur.io
+const createMailosaurEmail = randomName => 
+  `${randomName}.${Cypress.env('MAILOSAUR_SERVERID')}@mailosaur.io`;
+```
+
+<br/><br/>
+
+## (2) Explanation - what to test in an email and how
+
+First, let's elaborate on the setup we need.
+
+### Test Setup and Hybrid Approach: [Mailosaur Rest API](https://docs.mailosaur.com/reference) with `cy.request()` and [Mailosaur's Node package](https://www.npmjs.com/package/mailosaur) with `cy.task()`
+
+Mailosaur provides an [npm package](https://www.npmjs.com/package/mailosaur) and effectively all the Node code samples in the [api documentation](https://docs.mailosaur.com/reference) can be converted to `cy.task()`. Another approach is to implement Mailosaur's Rest API 'from scratch with `cy.request()`.
+
+While testing, we realized a combined approach where these utilities are chained into each other gave the best results.
+
+### Environment variables
+
+We recommend these values as environment variables. You can grab them from Mailosaur web application by creating a free trial account using any email. The trial account lasts for two weeks.
+
+```json
+  "MAILOSAUR_SERVERID": "******",
+  "MAILOSAUR_PASSWORD": "******",
+  "MAILOSAUR_API_KEY": "*******",
+  "MAILOSAUR_API": "https://mailosaur.com/api",
+  "MAILOSAUR_SERVERNAME": "any-name"
```

Comment:
this is not really "any-name" but a name generated by Mailosaur. I’m just thinking this might confuse some people.

Reply:
You might be right that it is generated, but it is customizable.

This is ours:
![image](https://user-images.githubusercontent.com/34237651/79250262-02ea3200-7e44-11ea-81f0-f66781d894c1.png)

Reply:
didn’t actually know that :) I’m pretty sure this wasn’t an option when I started using the service 🙂